# Customer Service Request Analysis
#### DESCRIPTION
You've been asked to perform data analysis of service request (311) calls from New York City. You've also been asked to utilize data wrangling techniques to understand the pattern in the data and visualize the major types of complaints.

#### **Perform the following steps:**
#### **1. Understand the dataset:**
   1. Identify the shape of the dataset
   2. Identify variables with null values

### **Import libraries**
- NumPy is the fundamental package for scientific computing in Python. 
- Pandas is a fast, powerful, flexible and easy to use Python data analysis and manipulation tool
- Matplotlib is a cross-platform, data visualization and graphical plotting library for Python
- Seaborn is a Python data visualization library based on matplotlib which is used for graphical statistical plotting

In [1]:
#import the required libraries
import numpy as pd
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# To view the plot in notebook
%matplotlib inline

### **Load Data**
Use the **`pd.read_csv`** function to read the *311 Service Request* CSV file into a pandas DataFrame

In [2]:
# Load the '311 Service Request' data set
df_311_dataset = pd.read_csv('311_Service_Requests_from_2010_to_Present.csv', low_memory=False)

In [3]:
# 1.1 Identify the shape of the dataset
df_311_dataset.shape

(364558, 53)

### **Explore the Data**
#### **1.1 Identify the shape of the dataset**
Use **`dataframe.shape`** to check the shape of the dataset

#### **Observations**:
There are **364,558 observations** (rows) and **53 variables** (columns) in the dataset

To further examine the dataset:  
Use the **`dataframe.head()`** function to view and examine the first 5 row of the data

In [11]:
# view the top 5 records
df_311_dataset.head()

Unique Key        Created Date         Closed Date Agency  \
0    32310363 2015-12-31 23:59:45 2016-01-01 00:55:15   NYPD   
1    32309934 2015-12-31 23:59:44 2016-01-01 01:26:57   NYPD   
2    32309159 2015-12-31 23:59:29 2016-01-01 04:51:03   NYPD   
3    32305098 2015-12-31 23:57:46 2016-01-01 07:43:13   NYPD   
4    32306529 2015-12-31 23:56:58 2016-01-01 03:24:42   NYPD   

                       Agency Name           Complaint Type  \
0  New York City Police Department  Noise - Street/Sidewalk   
1  New York City Police Department         Blocked Driveway   
2  New York City Police Department         Blocked Driveway   
3  New York City Police Department          Illegal Parking   
4  New York City Police Department          Illegal Parking   

                     Descriptor    Location Type Incident Zip  \
0              Loud Music/Party  Street/Sidewalk        10034   
1                     No Access  Street/Sidewalk        11105   
2                     No Access  Street/Sidewalk        10458   
3  Commercial Overnight Parking  Street/Sidewalk        10461   
4              Blocked Sidewalk  Street/Sidewalk        11373   

        Incident Address  ... Bridge Highway Name Bridge Highway Direction  \
0    71 VERMILYEA AVENUE  ...                 NaN                      NaN   
1        27-07 23 AVENUE  ...                 NaN                      NaN   
2  2897 VALENTINE AVENUE  ...                 NaN                      NaN   
3    2940 BAISLEY AVENUE  ...                 NaN                      NaN   
4          87-14 57 ROAD  ...                 NaN                      NaN   

  Road Ramp Bridge Highway Segment Garage Lot Name Ferry Direction  \
0       NaN                    NaN             NaN             NaN   
1       NaN                    NaN             NaN             NaN   
2       NaN                    NaN             NaN             NaN   
3       NaN                    NaN             NaN             NaN   
4       NaN                    NaN             NaN             NaN   

  Ferry Terminal Name   Latitude  Longitude  \
0                 NaN  40.865682 -73.923501   
1                 NaN  40.775945 -73.915094   
2                 NaN  40.870325 -73.888525   
3                 NaN  40.835994 -73.828379   
4                 NaN  40.733060 -73.874170   

                                   Location  
0   (40.86568153633767, -73.92350095571744)  
1  (40.775945312321085, -73.91509393898605)  
2  (40.870324522111424, -73.88852464418646)  
3   (40.83599404683083, -73.82837939584206)  
4  (40.733059618956815, -73.87416975810375)  

[5 rows x 53 columns]

#### **Observations**:
- Not all the columns are shown using the **`dataframe.head()`** function.
- There are definitely **NULL** (NaN) values in the dataset.
- The `Location` values are a tuple of (`Latitude`, `Longitude`)
- The `Incident Zip` which represents the Zip Code where the incident happens should be a categorical variable and not a floating point number.

Use the **`dataframe.info()`** to view the dataset in more detail - to get a concise summary of the dataframe.

In [5]:
# View the dataset info
df_311_dataset.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 364558 entries, 0 to 364557
Data columns (total 53 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Unique Key                      364558 non-null  int64  
 1   Created Date                    364558 non-null  object 
 2   Closed Date                     362177 non-null  object 
 3   Agency                          364558 non-null  object 
 4   Agency Name                     364558 non-null  object 
 5   Complaint Type                  364558 non-null  object 
 6   Descriptor                      358057 non-null  object 
 7   Location Type                   364425 non-null  object 
 8   Incident Zip                    361560 non-null  float64
 9   Incident Address                312859 non-null  object 
 10  Street Name                     312859 non-null  object 
 11  Cross Street 1                  307370 non-null  object 
 12  Cross Street 2  

#### **Observations**:
- We can see that the majority of our data columns are of dtype object (strings). The *`Unique Key`* is of dtype int64. The *`Incident Zip`*, *`X Coordinates (State Plane)`*, *`Y Coordinates (State Plane)`*, *`Latitude`*, and *`Longitude`* are of dtype float64.  
- The *`School or Citywide Complaint`*, *`Vehicle Type`*, *`Taxi Company Borough`*, *`Taxi Pick Up Location`*, and *`Garage Lot Name`* variables are also of dtype float64 dtype but have no (0) observations.
- The *`Ferry Direction`* and *`Ferry Terminal Name`* varaibles are of dtype object and have only 1 or 2 observations respectively.
- The *`Landmark`* variable is of dtype object and have only 375 observations.
- The *`Bridge Highway Name`*, *`Bridge Highway Direction`*, *`Road Ramp`*, and *`Bridge Highway Segment`* variables have less than 300 observations each.
- The above listed 12 variables that have lesser than 400 observations have more than 99.99% of missing (null) values so are not very useful features to use in modeling and will eventually be removed from the dataset. 
- The *`Incident Zip`* variable showing as dtype float64 should be reload as object (string) dtype or converted into a categorical variable since zip code values are not use in computation they should not be in numeric format.
- There are 4 date variables, *`Created Date`*, *`Closed Date`*, *`Due Date`*, and *`Resolution Action Updated Date`*, that are of dtype object. These date variables need to be loaded as or converted to datetime dtype.

#### Check the *`Incident Zip`* values to see how best to handle the conversion to categorical variable
- Use `dataframe.unique()` function to check the kind of values for the *`Incident Zip`* variable

In [6]:
df_311_dataset['Incident Zip'].unique()

array([10034., 11105., 10458., 10461., 11373., 11215., 10032., 10457.,
       11415., 11219., 11372., 10453., 11208., 11379., 11374., 11412.,
       11217., 11234., 10026., 10456., 10030., 10467., 11432., 10031.,
       11419., 10024., 11201., 11216., 10462.,    nan, 11385., 11414.,
       11213., 11375., 11211., 10312., 10017., 11417., 10002., 10027.,
       11209., 10035., 11418., 11421., 11205., 10468., 11355., 11358.,
       11210., 11368., 11427., 11436., 10308., 11364., 10011., 11423.,
       11230., 10003., 11221., 11416., 11378., 11236., 11218., 10029.,
       10028., 11214., 11207., 11369., 11223., 11220., 10302., 11420.,
       11354., 10473., 10301., 11103., 10465., 11377., 11212., 11365.,
       10472., 10452., 11203., 10469., 11237., 11434., 11101., 10460.,
       11229., 11206., 11102., 10466., 10009., 10033., 11694., 10022.,
       10470., 11433., 11428., 11413., 10463., 10471., 10474., 11228.,
       10014., 10475., 11225., 11233., 11370., 11204., 11435., 10459.,
      

#### **Observations:**
- Looks like other than the `NULL` (nan) values, all *`Incident Zip`* are valid *New York* zip codes so we can make sure that this variable is of categorical type by specifying that the *`Incident Zip`* value should be of dtype string when reloading the dataset. Alternatively, we can use the `Dataframe[colName].astype('str')` function to convert the zip code from numerical to categorical.

#### Convert date variables from object type to datetime type and ensure that *`Incident Zip`* is a categorical variable
- Reload the data from the *311 Service Requests* CSV file with the `parse_dates` parameter listing the date variables to be parsed in a list and specify the dtype for *`Incident Zip`* as dtype `str`. 
- View the dataset info using the `dataframe.info()` function to validate the change in the data type (dtype).

**Note:** It is easier to reload the data with the `parse_dates` parameter here. Alternatively, the dtype for the date variables can be converted using `pd.to_datetime()` function, for example, `df[dateColumn] = pd.to_datetime(df[dateColumn])`.

In [7]:
# Reload the '311 Service Request' data set with parse_dates parameter
df_311_dataset = pd.read_csv('311_Service_Requests_from_2010_to_Present.csv', low_memory=False, \
                             parse_dates=['Created Date', 'Closed Date', 'Due Date', 'Resolution Action Updated Date'], \
                             dtype={'Incident Zip':'str'})

# View the dataset info
df_311_dataset.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 364558 entries, 0 to 364557
Data columns (total 53 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   Unique Key                      364558 non-null  int64         
 1   Created Date                    364558 non-null  datetime64[ns]
 2   Closed Date                     362177 non-null  datetime64[ns]
 3   Agency                          364558 non-null  object        
 4   Agency Name                     364558 non-null  object        
 5   Complaint Type                  364558 non-null  object        
 6   Descriptor                      358057 non-null  object        
 7   Location Type                   364425 non-null  object        
 8   Incident Zip                    361560 non-null  object        
 9   Incident Address                312859 non-null  object        
 10  Street Name                     312859 non-null  object 

#### **Observations**:
- The data type for the 4 date variables had been parsed and stored as datetime64[ns] dtype in the dataFrame.
- The *`Incident Zip`* variable had also been captured as dtype object.

#### **1.2 Identify variables with null values**
Use the `isna().sum(axis=0)` function to check the number of NULL values in the DataFrame. The `isna()` function returns a DataFrame of Boolean values that are True for null values. The `sum()` function returns the total number of null values for each of the variables. The attribute `axis=0` indicates that the rows values were summed up. 

In [8]:
# 1.2 Identify variables with null values
df_311_dataset.isna().sum(axis=0)

Unique Key                             0
Created Date                           0
Closed Date                         2381
Agency                                 0
Agency Name                            0
Complaint Type                         0
Descriptor                          6501
Location Type                        133
Incident Zip                        2998
Incident Address                   51699
Street Name                        51699
Cross Street 1                     57188
Cross Street 2                     57805
Intersection Street 1             313438
Intersection Street 2             314046
Address Type                        3252
City                                2997
Landmark                          364183
Facility Type                       2389
Status                                 0
Due Date                               3
Resolution Description                 0
Resolution Action Updated Date      2402
Community Board                        0
Borough         

#### **Observations:**
- Almost two third of the variables in the dataset contain null values. 

To check the ratio or percentage of null values in each of the variables, use the `dataframe.isnull().mean()` function to find the null values mean for each variable and use the `.sort_values()` function to sort the result in descending order (using attribute `ascending=False`).

In [9]:
df_311_dataset.isnull().mean().sort_values(ascending=False)

School or Citywide Complaint      1.000000
Vehicle Type                      1.000000
Taxi Company Borough              1.000000
Taxi Pick Up Location             1.000000
Garage Lot Name                   1.000000
Ferry Direction                   0.999997
Ferry Terminal Name               0.999995
Road Ramp                         0.999281
Bridge Highway Segment            0.999281
Bridge Highway Name               0.999185
Bridge Highway Direction          0.999185
Landmark                          0.998971
Intersection Street 2             0.861443
Intersection Street 1             0.859775
Cross Street 2                    0.158562
Cross Street 1                    0.156869
Street Name                       0.141813
Incident Address                  0.141813
Descriptor                        0.017833
X Coordinate (State Plane)        0.011054
Latitude                          0.011054
Longitude                         0.011054
Y Coordinate (State Plane)        0.011054
Location   

#### **Observations:**
- 14 variables have more than 80% of null values in its observations. Since these variables have hardly any useful data with so many missing values, they can be removed from the dataset.  
- The *`Cross Street 1`*, *`Cross Street 2`*, *`Street Name`*, and *`Incident Address`* variables have around 14 ~ 16% missing (null) values.  
- The *`Descriptor`* variable has approximately 1.8% of missing values in its observations.  
- The *`X Coordinate (State Plane)`*, *`Y Coordinate (State Plane)`*, *`Latitude`*, *`Longitude`*, and *`Location`* variables have approximately 1.1% of missing values in its observations.   
- The *`Address Type`*, *`Incident Zip`*, and *`City`* have about 0.8% of missing values in its observations.  
- The *`Facility Type`*, *`Closed Date`* and *`Resolution Action Date`* variables have about 0.6% of missing values.  
- The *`Location Type`* variable has 0.0365% of missing values  
- The *`Due Date`*, *`School Region`*, *`School Code`*, and *`School Zip`* variables have less than or equal to 0.0008% of missing values.

#### Get the column labels for those variables that have null values
Use the `dataframe.isnull().any()` functions to get the Series of variables with null values indicators and `dataframe.columns` to get the label for these variables (columns) that have null values.

In [10]:
# Display the variables (column names) with null values
columns_with_null_values = df_311_dataset.columns[df_311_dataset.isnull().any()]
print(columns_with_null_values, '\n')
print('Number of variables with null values: ', len(columns_with_null_values))

Index(['Closed Date', 'Descriptor', 'Location Type', 'Incident Zip',
       'Incident Address', 'Street Name', 'Cross Street 1', 'Cross Street 2',
       'Intersection Street 1', 'Intersection Street 2', 'Address Type',
       'City', 'Landmark', 'Facility Type', 'Due Date',
       'Resolution Action Updated Date', 'X Coordinate (State Plane)',
       'Y Coordinate (State Plane)', 'School Region', 'School Code',
       'School Zip', 'School or Citywide Complaint', 'Vehicle Type',
       'Taxi Company Borough', 'Taxi Pick Up Location', 'Bridge Highway Name',
       'Bridge Highway Direction', 'Road Ramp', 'Bridge Highway Segment',
       'Garage Lot Name', 'Ferry Direction', 'Ferry Terminal Name', 'Latitude',
       'Longitude', 'Location'],
      dtype='object') 

Number of variables with null values:  35


#### **Observations:**
35 variables had been identified with null values. These variables are:
1. 'Closed Date'                 
2. 'Descriptor'
3. 'Location Type'
4. 'Incident Zip'
5. 'Incident Address'
6. 'Street Name'
7. 'Cross Street 1'
8. 'Cross Street 2'
9. 'Intersection Street 1'
10. 'Intersection Street 2'
11. 'Address Type'
12. 'City'
13.  'Landmark'
14. 'Facility Type'
15. 'Due Date'
16. 'Resolution Action Updated Date'
17. 'X Coordinate (State Plane)'
18. 'Y Coordinate (State Plane)'
19. 'School Region'
20. 'School Code'
21. 'School Zip'
22. 'School or Citywide Complaint'
23. 'Vehicle Type'
24. 'Taxi Company Borough'
25. 'Taxi Pick Up Location'
26. 'Bridge Highway Name'
27. 'Bridge Highway Direction'
28. 'Road Ramp'
29. 'Bridge Highway Segment'
30. 'Garage Lot Name'
31. 'Ferry Direction'
32. 'Ferry Terminal Name'
33. 'Latitude'
34. 'Longitude'
35. 'Location'

#### **2. Perform basic data exploratory analysis:**
   1. Utilize missing value treatment
   2. Analyze the date column and remove the entries if it has an incorrect timeline
   3. Draw a frequency plot for city-wise complaints
   4. Draw scatter and hexbin plots for complaint concentration across Brooklyn
   
#### **2.1 Utilize missing value treatment**
#### Remove variables (columns) having more than 80% missing (null) values from the dataset
Since the 14 variables that have more than 80% of missing values are not very useful, we can remove (drop) the columns.

- Filter data based on the columns with the condition that only variables with less than 80% missing values in its observations are kept and stored in a new DataFrame, **df_311_data**.
- Then use the `dataframe.info()` function to validate that the columns with greater than 80% missing values were removed.

In [14]:
# Filter data based on the columns that have less than 80% of missing values in its observations into a new DataFrame, df_311_data.
df_311_data = df_311_dataset.loc[:, df_311_dataset.isnull().mean() < 0.8]
df_311_data.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 364558 entries, 0 to 364557
Data columns (total 39 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   Unique Key                      364558 non-null  int64         
 1   Created Date                    364558 non-null  datetime64[ns]
 2   Closed Date                     362177 non-null  datetime64[ns]
 3   Agency                          364558 non-null  object        
 4   Agency Name                     364558 non-null  object        
 5   Complaint Type                  364558 non-null  object        
 6   Descriptor                      358057 non-null  object        
 7   Location Type                   364425 non-null  object        
 8   Incident Zip                    361560 non-null  object        
 9   Incident Address                312859 non-null  object        
 10  Street Name                     312859 non-null  object 

In [15]:
# Check the dimension of the filtered dataset
df_311_data.shape

(364558, 39)

#### **Observations:**
Using the `dataframe.info()` and `dataframe.shape` functions, it can be verified that 14 columns were dropped from the dataset 

#### Perform Basic Data Analysis on the dataset
Let's take a look at the remaining data in the dataset to see what unique values they have. Use the `dataframe.value_counts()` function with parameter `dropna=False` to include the NULL value in the unique value counts and the `dataframe.count()` function to count the number of unique values for each variable.

In [16]:
print('Number of unique values for each of the variables are as follows: ')
print(' 0. Unique Key: ', '\t\t\t', df_311_data['Unique Key'].value_counts(dropna=False).count())
print(' 1. Created Date: ', '\t\t\t', df_311_data['Created Date'].value_counts(dropna=False).count())
print(' 2. Closed Date: ', '\t\t\t', df_311_data['Closed Date'].value_counts(dropna=False).count())
print(' 3. Agency: ', '\t\t\t\t', df_311_data['Agency'].value_counts(dropna=False).count())
print(' 4. Agency Name: ', '\t\t\t', df_311_data['Agency Name'].value_counts(dropna=False).count())
print(' 5. Complaint Type: ', '\t\t\t', df_311_data['Complaint Type'].value_counts(dropna=False).count())
print(' 6. Descriptor: ', '\t\t\t', df_311_data['Descriptor'].value_counts(dropna=False).count())
print(' 7. Location Type: ', '\t\t\t', df_311_data['Location Type'].value_counts(dropna=False).count())
print(' 8. Incident Zip: ', '\t\t\t', df_311_data['Incident Zip'].value_counts(dropna=False).count())
print(' 9. Incident Address: ', '\t\t\t', df_311_data['Incident Address'].value_counts(dropna=False).count())
print('10. Street Name: ', '\t\t\t', df_311_data['Street Name'].value_counts(dropna=False).count())
print('11. Cross Street 1: ', '\t\t\t', df_311_data['Cross Street 1'].value_counts(dropna=False).count())
print('12. Cross Street 2: ', '\t\t\t', df_311_data['Cross Street 2'].value_counts(dropna=False).count())
print('13. Address Type: ', '\t\t\t', df_311_data['Address Type'].value_counts(dropna=False).count())
print('14. City: ', '\t\t\t\t', df_311_data['City'].value_counts(dropna=False).count())
print('15. Facility Type: ', '\t\t\t', df_311_data['Facility Type'].value_counts(dropna=False).count())
print('16. Status: ', '\t\t\t\t', df_311_data['Status'].value_counts(dropna=False).count())
print('17. Due Date: ', '\t\t\t\t', df_311_data['Due Date'].value_counts(dropna=False).count())
print('18. Resolution Description: ', '\t\t', df_311_data['Resolution Description'].value_counts(dropna=False).count())
print('19. Resolution Action Updated Date: ', '\t', df_311_data['Resolution Action Updated Date'].value_counts(dropna=False).count())
print('20. Community Board: ', '\t\t\t', df_311_data['Community Board'].value_counts(dropna=False).count())
print('21. Borough: ', '\t\t\t\t', df_311_data['Borough'].value_counts(dropna=False).count())
print('22. X Coordinate (State Plane): ', '\t', df_311_data['X Coordinate (State Plane)'].value_counts(dropna=False).count())
print('23. Y Coordinate (State Plane): ', '\t', df_311_data['Y Coordinate (State Plane)'].value_counts(dropna=False).count())
print('24. Park Facility Name: ', '\t\t', df_311_data['Park Facility Name'].value_counts(dropna=False).count())
print('25. Park Borough: ', '\t\t\t', df_311_data['Park Borough'].value_counts(dropna=False).count())
print('26. School Name: ', '\t\t\t', df_311_data['School Name'].value_counts(dropna=False).count())
print('27. School Number: ', '\t\t\t', df_311_data['School Number'].value_counts(dropna=False).count())
print('28. School Region: ', '\t\t\t', df_311_data['School Region'].value_counts(dropna=False).count())
print('29. School Code: ', '\t\t\t', df_311_data['School Code'].value_counts(dropna=False).count())
print('30. School Phone Number: ', '\t\t', df_311_data['School Phone Number'].value_counts(dropna=False).count())
print('31. School Address: ', '\t\t\t', df_311_data['School Address'].value_counts(dropna=False).count())
print('32. School City: ', '\t\t\t', df_311_data['School City'].value_counts(dropna=False).count())
print('33. School State: ', '\t\t\t', df_311_data['School State'].value_counts(dropna=False).count())
print('34. School Zip: ', '\t\t\t', df_311_data['School Zip'].value_counts(dropna=False).count())
print('35. School Not Found: ', '\t\t\t', df_311_data['School Not Found'].value_counts(dropna=False).count())
print('36. Latitude: ', '\t\t\t\t', df_311_data['Latitude'].value_counts(dropna=False).count())
print('37. Longitude: ', '\t\t\t', df_311_data['Longitude'].value_counts(dropna=False).count())
print('38. Location: ', '\t\t\t\t', df_311_data['Location'].value_counts(dropna=False).count())

Number of unique values for each of the variables are as follows: 
 0. Unique Key:  			 364558
 1. Created Date:  			 362018
 2. Closed Date:  			 339838
 3. Agency:  				 1
 4. Agency Name:  			 3
 5. Complaint Type:  			 24
 6. Descriptor:  			 46
 7. Location Type:  			 19
 8. Incident Zip:  			 202
 9. Incident Address:  			 126373
10. Street Name:  			 7694
11. Cross Street 1:  			 6235
12. Cross Street 2:  			 6065
13. Address Type:  			 6
14. City:  				 54
15. Facility Type:  			 2
16. Status:  				 4
17. Due Date:  				 362016
18. Resolution Description:  		 18
19. Resolution Action Updated Date:  	 340834
20. Community Board:  			 75
21. Borough:  				 6
22. X Coordinate (State Plane):  	 68411
23. Y Coordinate (State Plane):  	 79925
24. Park Facility Name:  		 2
25. Park Borough:  			 6
26. School Name:  			 2
27. School Number:  			 2
28. School Region:  			 2
29. School Code:  			 2
30. School Phone Number:  		 2
31. School Address:  			 2
32. School City:  			 2
33. Scho

#### **Observations:**
- The *`Unique Key`* has 364558 unique values which is the same as the total number of observations in the dataset - confirmed as unique.
- There is only 1 unique *`Agency`* but 3 unique *`Agency Name`*
- Both the *`Borough`* and *`Park Borough`* variables have 6 unique values with no null values. The unique values are probably the same.
- The number of unique *`City`* variable is 54 with 0.8221% of null values
- Other than the *`Agency`* variable, there are 12 other variables that only have 1 or 2 unique values. We can take a look at the values.

Let's take a look at the variables that have less than 10 unique values.

In [17]:
print(df_311_data['Agency'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['Agency Name'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['Address Type'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['Facility Type'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['Status'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['Borough'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['Park Borough'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['Park Facility Name'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School Name'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School Number'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School Region'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School Code'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School Phone Number'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School Address'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School City'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School State'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School Zip'].value_counts(dropna=False), '\n-----------------------------------------------------------')
print(df_311_data['School Not Found'].value_counts(dropna=False))

NYPD    364558
Name: Agency, dtype: int64 
-----------------------------------------------------------
New York City Police Department    364548
Internal Affairs Bureau                 8
NYPD                                    2
Name: Agency Name, dtype: int64 
-----------------------------------------------------------
ADDRESS         292699
INTERSECTION     50514
BLOCKFACE        13786
LATLONG           3931
NaN               3252
PLACENAME          376
Name: Address Type, dtype: int64 
-----------------------------------------------------------
Precinct    362169
NaN           2389
Name: Facility Type, dtype: int64 
-----------------------------------------------------------
Closed      362114
Open          1640
Assigned       802
Draft            2
Name: Status, dtype: int64 
-----------------------------------------------------------
BROOKLYN         118864
QUEENS           100766
MANHATTAN         77462
BRONX             49169
STATEN ISLAND     15339
Unspecified        2958
Name:

#### **Observations:**
- The *`Agency`* variable only have 1 unique value *NYPD* which shows that this dataset is specific for *New York Police Department* 311 Service Request.
- *New York City Police Department* and *NYPD* should be the same *`Agency Name`*. Looks like only 8 complaints goes to *Internal Affairs Bureau*.
- There seems to have only 1 unique value for *`Facility Type`* which is *Precinct*. The missing values for this variable can probably be filled with this value.
- The *`Borough`* and *`Park Borough`* variables seem to have the same information. Although there is no missing values in these 2 variables, there are 2958 with the value of *Unspecified*. Looks like the missing values for these variables are filled with this value.
- The *`Park Facility Name`*, *`School Name`*, *`School Number`*, *`School Region`*, *`School Code`*, *`School Phone Number`*, *`School Address`*, *`School City`*, *`School State`*, and *`School Zip`* variables only have 1 observation. The other 364557 observations in these variables also have the value of *Unspecified*. 
- The missing values in the *`Address Type`* variable probably can be filled with the *Unspecified* value as well. 
- In that case, to be consistent, probably all the categorical variables that have missing values, except for *`Facitlity Type`*, can be filled with the value *Unspecified*.
- All the 364558 observations for the *`School Not Found`* variable has the value of *N*.

#### Drop redundant columns
Since the *`Park Facility Name`*, *`School Name`*, *`School Number`*, *`School Region`*, *`School Code`*, *`School Phone Number`*, *`School Address`*, *`School City`*, *`School State`*, and *`School Zip`* variables have 364557 observations with *Unspecified* value and only 1 observation with either null or valid value, these variables are basically not very useful for modeling. So, let's just drop these columns from the dataset and save the result to a new DataFrame, **df_311_final**.

In [28]:
# Drop the specified columns (axis=1)
df_311_final = df_311_data.drop(['Park Facility Name', 'School Name', 'School Number', 'School Region', 'School Code', 'School Phone Number', 'School Address', 'School City', 'School State', 'School Zip'], axis=1)

# Validate that the 10 columns had been droped in the new DataFrame 
print(df_311_final.info(memory_usage='deep'))
df_311_final.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 364558 entries, 0 to 364557
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   Unique Key                      364558 non-null  int64         
 1   Created Date                    364558 non-null  datetime64[ns]
 2   Closed Date                     362177 non-null  datetime64[ns]
 3   Agency                          364558 non-null  object        
 4   Agency Name                     364558 non-null  object        
 5   Complaint Type                  364558 non-null  object        
 6   Descriptor                      358057 non-null  object        
 7   Location Type                   364425 non-null  object        
 8   Incident Zip                    361560 non-null  object        
 9   Incident Address                312859 non-null  object        
 10  Street Name                     312859 non-null  object 

Unique Key        Created Date         Closed Date Agency  \
0    32310363 2015-12-31 23:59:45 2016-01-01 00:55:15   NYPD   
1    32309934 2015-12-31 23:59:44 2016-01-01 01:26:57   NYPD   

                       Agency Name           Complaint Type        Descriptor  \
0  New York City Police Department  Noise - Street/Sidewalk  Loud Music/Party   
1  New York City Police Department         Blocked Driveway         No Access   

     Location Type Incident Zip     Incident Address  ...  \
0  Street/Sidewalk        10034  71 VERMILYEA AVENUE  ...   
1  Street/Sidewalk        11105      27-07 23 AVENUE  ...   

  Resolution Action Updated Date Community Board    Borough  \
0            2016-01-01 00:55:15    12 MANHATTAN  MANHATTAN   
1            2016-01-01 01:26:57       01 QUEENS     QUEENS   

  X Coordinate (State Plane) Y Coordinate (State Plane) Park Borough  \
0                  1005409.0                   254678.0    MANHATTAN   
1                  1007766.0                   221986.0       QUEENS   

  School Not Found   Latitude  Longitude  \
0                N  40.865682 -73.923501   
1                N  40.775945 -73.915094   

                                   Location  
0   (40.86568153633767, -73.92350095571744)  
1  (40.775945312321085, -73.91509393898605)  

[2 rows x 29 columns]

#### Observations:
- The 10 variables (columns) had been dropped.
- There are now only 29 variables in the final dataset.

#### Dealing with Missing Values
There are 3 types of data with missing values in the dataset, namely, numerical, categorical and datetime.
- To be consistent, all categorical variables with missing (null) values , except for the *`Facility Type`* variable, can be filled with the *Unspecified* value to create a new category.
- The *`Facility Type`* variable has only 1 unique value, *Precinct*, for all except the missing value observations, this unique value can be used to replace the missing values for this variable.
- The numerical variables with missing values are *`X Coordinate (State Plane)`*, *`Y Coordinate (State Plane)`*, *`Latitude`* and *`Longitude`*. Since the values for these variables need to be valid values pertain to the location of the incident and are not used for computation, it should not be replaced with any values and be kept as null values.
- The datetime variables with missing values, *`Closed Date`*, *`Due Date`*, and *`Resolution Action Updated Date`* could be due to the incident had not been closed or have no updates. So, theses missing (null) values will also be kept as null values.

Therefore, only the categorical variables with missing values will be treated.

In [29]:
# Replace the categorical variables with the 'Unspecifed' value
df_311_final = df_311_final.fillna(value = {'Descriptor':'Unspecified',
                                           'Location Type':'Unspecified',
                                           'Incident Zip':'Unspecified',
                                           'Incident Address':'Unspecified',
                                           'Street Name':'Unspecified',
                                           'Cross Street 1':'Unspecified',
                                           'Cross Street 2':'Unspecified',
                                           'Address Type':'Unspecified',
                                           'City':'Unspecified',
                                           'Facility Type':'Precinct',
                                           'Location':'Unspecified'})

# Check to make sure the numerical and datetime values still have null values
df_311_final.isna().sum(axis=0)

Unique Key                           0
Created Date                         0
Closed Date                       2381
Agency                               0
Agency Name                          0
Complaint Type                       0
Descriptor                           0
Location Type                        0
Incident Zip                         0
Incident Address                     0
Street Name                          0
Cross Street 1                       0
Cross Street 2                       0
Address Type                         0
City                                 0
Facility Type                        0
Status                               0
Due Date                             3
Resolution Description               0
Resolution Action Updated Date    2402
Community Board                      0
Borough                              0
X Coordinate (State Plane)        4030
Y Coordinate (State Plane)        4030
Park Borough                         0
School Not Found         

#### **Observations**
- The missing values in the 3 date variables, *`Closed Date`*, *`Due Date`*, *`Resolution Action Updated Date`*, and the 4 numerical variables, *`X Coordinate (State Plane)`*, *`Y Coordinate (State Plane)`*, *`Latitude`*, *`Longitude`* had not been replaced as planned.
- All the categorical variables no longer have any missing values.

#### **2.2 Analyze the date column and remove the entries if it has an incorrect timeline**
- The incident *`Created Date`* should be earlier than the *`Closed Date`*, *`Due Date`*, or *`Resolution Action Updated Date`*.
- Check each of the other date variables against *`Created Date`* to ensure the timeline is correct.

First, check if any *`Created Date`* has a later date value than the *`Closed Date`*.

In [30]:
# 2.2 Analyze the data column and remove the entries if it has an incorrect timeline
# First, check if any 'Created Date' has a later date value than the 'Closed Date'
df_311_final[df_311_final['Created Date'] > df_311_final['Closed Date']]

Empty DataFrame
Columns: [Unique Key, Created Date, Closed Date, Agency, Agency Name, Complaint Type, Descriptor, Location Type, Incident Zip, Incident Address, Street Name, Cross Street 1, Cross Street 2, Address Type, City, Facility Type, Status, Due Date, Resolution Description, Resolution Action Updated Date, Community Board, Borough, X Coordinate (State Plane), Y Coordinate (State Plane), Park Borough, School Not Found, Latitude, Longitude, Location]
Index: []

[0 rows x 29 columns]

#### **Observations:**
There were no incorrect timeline between *`Created Date`* and *`Closed Date`*.

Then, check if any *`Created Date`* has a later date value than the *`Due Date`*.

In [31]:
df_311_final[df_311_final['Created Date'] > df_311_final['Due Date']]

Empty DataFrame
Columns: [Unique Key, Created Date, Closed Date, Agency, Agency Name, Complaint Type, Descriptor, Location Type, Incident Zip, Incident Address, Street Name, Cross Street 1, Cross Street 2, Address Type, City, Facility Type, Status, Due Date, Resolution Description, Resolution Action Updated Date, Community Board, Borough, X Coordinate (State Plane), Y Coordinate (State Plane), Park Borough, School Not Found, Latitude, Longitude, Location]
Index: []

[0 rows x 29 columns]

#### **Observations:**
There were no incorrect timeline between *`Created Date`* and *`Due Date`*.

Finally, check if any *`Created Date`* has a later date value than the *`Resolution Action Updated Date`*.

In [32]:
df_311_final[df_311_final['Created Date'] > df_311_final['Resolution Action Updated Date']]

Unique Key        Created Date Closed Date Agency  \
61086    31880851 2015-11-01 01:57:24         NaT   NYPD   
61095    31886638 2015-11-01 01:54:37         NaT   NYPD   
61098    31884397 2015-11-01 01:53:10         NaT   NYPD   
61111    31884894 2015-11-01 01:46:36         NaT   NYPD   
61119    31885698 2015-11-01 01:43:18         NaT   NYPD   

                           Agency Name           Complaint Type  \
61086  New York City Police Department       Noise - Commercial   
61095  New York City Police Department         Blocked Driveway   
61098  New York City Police Department             Noise - Park   
61111  New York City Police Department  Noise - Street/Sidewalk   
61119  New York City Police Department  Noise - Street/Sidewalk   

             Descriptor        Location Type Incident Zip  \
61086  Loud Music/Party  Club/Bar/Restaurant        11378   
61095         No Access      Street/Sidewalk        11233   
61098  Loud Music/Party      Park/Playground        10024   
61111      Loud Talking      Street/Sidewalk        10032   
61119  Loud Music/Party      Street/Sidewalk        10032   

          Incident Address  ... Resolution Action Updated Date  \
61086       53-73 65 PLACE  ...            2015-11-01 01:00:44   
61095  847 HERKIMER STREET  ...            2015-11-01 01:02:57   
61098          Unspecified  ...            2015-11-01 01:33:17   
61111  643 WEST 172 STREET  ...            2015-11-01 01:22:04   
61119    WEST 172ND STREET  ...            2015-11-01 01:22:06   

      Community Board    Borough X Coordinate (State Plane)  \
61086       05 QUEENS     QUEENS                  1011904.0   
61095     03 BROOKLYN   BROOKLYN                  1005030.0   
61098    07 MANHATTAN  MANHATTAN                   992474.0   
61111    12 MANHATTAN  MANHATTAN                  1000946.0   
61119    01 MANHATTAN  MANHATTAN                  1001119.0   

      Y Coordinate (State Plane) Park Borough School Not Found   Latitude  \
61086                   204981.0       QUEENS                N  40.729259   
61095                   186412.0     BROOKLYN                N  40.678310   
61098                   226751.0    MANHATTAN                N  40.789051   
61111                   246854.0    MANHATTAN                N  40.844217   
61119                   246761.0    MANHATTAN                N  40.843961   

       Longitude                                 Location  
61086 -73.900223  (40.72925894967772, -73.90022336472336)  
61095 -73.925082  (40.678310248732046, -73.9250822918449)  
61098 -73.970301    (40.78905142646646, -73.970300867221)  
61111 -73.939656  (40.84421656446944, -73.93965613439441)  
61119 -73.939031  (40.843960977988715, -73.9390310996927)  

[5 rows x 29 columns]

#### **Observations:**
There were 5 observations with incorrect timeline between *`Created Date`* and *`Resolution Action Updated Date`*.

Remove these 5 rows with incorrect timeline.

In [33]:
# Remove the 5 rows and keep only the observations with 'Created Date' earlier than the 'Resolution Action Updated Date'
df_311_final = df_311_final.drop(df_311_final[df_311_final['Created Date'] > df_311_final['Resolution Action Updated Date']].index)

# Check if the 5 rows had been removed using the df.shape function
print(df_311_final.shape)

# Check if there are still observations with incorrect timeline
df_311_final[df_311_final['Created Date'] > df_311_final['Resolution Action Updated Date']]

(364553, 29)


Empty DataFrame
Columns: [Unique Key, Created Date, Closed Date, Agency, Agency Name, Complaint Type, Descriptor, Location Type, Incident Zip, Incident Address, Street Name, Cross Street 1, Cross Street 2, Address Type, City, Facility Type, Status, Due Date, Resolution Description, Resolution Action Updated Date, Community Board, Borough, X Coordinate (State Plane), Y Coordinate (State Plane), Park Borough, School Not Found, Latitude, Longitude, Location]
Index: []

[0 rows x 29 columns]

#### **Observations:**
- The number of observations in the final dataset is now 364,553 showing that 5 rows had been removed from the original 364,558 observations.
- Now there is no incorrect timeline between the `Created Date` and the `Resolution Action Updated Date`.

#### **2.3 Draw a frequency plot for city-wise complaints**

In [ ]:
# 2.3. Draw a frequency plot for city-wise complaints

#### **Observations:**


#### **2.4 Draw scatter and hexbin plots for complaint concentration across Brooklyn**

In [ ]:
# 2.4 Draw scatter and hexbin plots for complaint concentration across Brooklyn

#### **Observations:**

##### **3. Find major types of complaints:**
   1. Plot a bar graph of count vs. complaint types
   2. Find the top 10 types of complaints
   3. Display the types of complaints in each city in a separate dataset
   
#### **3.1 Plot a bar graph of count vs. complaint types**

In [ ]:
# 3.1 Plot a bar graph of count vs. complaint types

#### **Observations:**

#### **3.2 Find the top 10 types of complaints**

In [35]:
# 3.2 Find the top 10 types of complaints

#### **Observations:**

#### **3.3 Display the types of complaints in each city in a separate dataset**


In [36]:
# 3.3 Display the types of complaints in each city in a separate datase

#### **Observations:**

##### **4. Visualize the major types of complaints in each city**


In [37]:
# 4. Visualize the major types of complaints in each city

##### **5. Check if the average response time across various types of complaints**

In [38]:
# 5. Check if the average response time across various types of complaints

##### **6. Identify significant variables by performing a statistical analysis using p-values and chi-square values (Optional)**

In [39]:
# 6. Identify significant variables by performing a statistical analysis using p-values and chi-square values

ERROR: Could not find a version that satisfies the requirement xelatex (from versions: none)
ERROR: No matching distribution found for xelatex
